src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1154 @@
+            } else {
+                $cond_satisfied = 0 unless exists($config_options->{$cond_option_set});
+            }
+        }
+        if (!$cond_satisfied) {
+            # When the condition is not satisfied and if option remove_if_undef is set,
+            # remove configuration line (if present).
+            *$self->{LOG}->debug(1, "$function_name: condition not satisfied, flag set to remove matching configuration lines");
+            $rule_info{remove_matching_lines} = 1;
+            return \%rule_info;
+        }
+    }
+
+    my @option_sets;
+    ($rule_info{attribute}, my $option_sets_str) = split /:/, $rule;
+    if ($option_sets_str) {
+        @option_sets = split /\s*,\s*/, $option_sets_str;
+    }
+    $rule_info{option_sets} = \@option_sets;
+
+    return \%rule_info;
+}
+
+
+=pod
+
+=item _apply_rules
+
+Apply configuration rules. This method is the real workhorse of the rule-based editor.
+
+Arguments :
+    config_rules: config rules corresponding to the file to build
+    config_options: configuration parameters used to build actual configuration
+    parser_options: a hash setting options to modify the behaviour of this function
+
+Supported entries for options hash:
+    always_rules_only: if true, apply only rules with ALWAYS condition (D: false)
+    remove_if_undef: if true, remove matching configuration line is rule condition is not met (D: false)
+
+=cut
+
+sub _apply_rules
+{
+    my $function_name = "_apply_rules";
+    my ($self, $config_rules, $config_options, $parser_options) = @_;
+
+    unless ($config_rules) {
+        *$self->{LOG}->error("$function_name: 'config_rules' argument missing (internal error)");
+        return 1;
+    }
+    unless ($config_options) {
+        *$self->{LOG}->error("$function_name: 'config_options' argument missing (internal error)");
+        return 1;
+    }
+    unless (defined($parser_options)) {
+        *$self->{LOG}->debug(2, "$function_name: 'parser_options' undefined");
+        $parser_options = {};
+    }
+    if (defined($parser_options->{remove_if_undef})) {
+        *$self->{LOG}->debug(1, "$function_name: 'remove_if_undef' option set to " . $parser_options->{remove_if_undef});
+    } else {
+        *$self->{LOG}->debug(1, "$function_name: 'remove_if_undef' option not defined: assuming $LINE_OPT_DEF_REMOVE_IF_UNDEF");
+        $parser_options->{remove_if_undef} = $LINE_OPT_DEF_REMOVE_IF_UNDEF;
+    }
+
+
+    # Loop over all config rule entries, sorted by keyword alphabetical order.
+    # Config rules are stored in a hash whose key is the variable to write
+    # and whose value is the rule itself.
+    # If the variable name start with a '-', this means that the matching configuration
+    # line must be commented out unconditionally.
+    # Each rule format is '[condition->]attribute:option_set[,option_set,...];line_fmt' where
+    #     condition: either a role that must be enabled or ALWAYS if the rule must be applied
```

Comment:
The use of the word role here is confusing. It's very close to "rule" so could easily be a typographic error, and it's not clear what it means given that the word role appears only in this small section and no other part of the POD.